File: HW UIView/HW UIView/ViewController.m
//  ViewController.m
//  HW UIView
//
//  Created by Kirill Fedorov on 28/02/2019.
//  Copyright © 2019 Kirill Fedorov. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@property (assign, nonatomic) CGFloat cellSize;
@property (strong, nonatomic) NSMutableArray *checkersArray;
@property (strong, nonatomic) UIView *field;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
//    1. В цикле создавайте квадратные UIView с черным фоном и расположите их в виде шахматной доски
//    2. доска должна иметь столько клеток, как и настоящая шахматная

    self.checkersArray = [NSMutableArray array];
    CGFloat width = self.view.bounds.size.width;
    CGFloat height = self.view.bounds.size.height;
    CGFloat minSide = MIN(width, height);
    NSInteger cellsInLine = 20;
    
    self.field = [[UIView alloc] initWithFrame:CGRectMake(self.view.center.x - minSide / 2, self.view.center.y - minSide / 2, minSide, minSide)];
    [self.view addSubview:self.field];
    self.field.backgroundColor = [UIColor redColor];
    
    self.cellSize = MIN(width, height) / cellsInLine;
    CGFloat multiplier = 0.7;
    CGFloat checkerSize = self.cellSize * multiplier;
    CGFloat indentInCell = self.cellSize * (1 - multiplier) / 2;
    NSInteger linesWithChekers = 16;
    NSInteger cellsWithOutChekers = cellsInLine - linesWithChekers / 2;
    
    for (NSInteger i = 0; i < cellsInLine; i++) {
        for (NSInteger j = 0; j < cellsInLine; j++) {
            BOOL isEven = (i + j) % 2 == 0;
            
            UIView *cell = [[UIView alloc]initWithFrame:CGRectMake(self.cellSize * j, self.cellSize * i, self.cellSize, self.cellSize)];
            
            if (!isEven) {
                cell.backgroundColor = [UIColor lightGrayColor];
            } else {
                cell.backgroundColor = [UIColor blackColor];
            }
            [self.field addSubview:cell];
            
            UIView *checker = [[UIView alloc] initWithFrame:CGRectMake(self.cellSize * j + indentInCell,
                                                                          self.cellSize * i + indentInCell,
                                                                          checkerSize,
                                                                          checkerSize)];

            if (i < cellsInLine - cellsWithOutChekers && isEven) {
                checker.backgroundColor = [UIColor whiteColor];
            }
            if (i >= cellsWithOutChekers && isEven) {
                checker.backgroundColor = [UIColor redColor];
            }
            [self.field addSubview:checker];
            [self.checkersArray addObject:checker];
        }
    }
    
    
//    3. Доска должна быть вписана в максимально возможный квадрат, т.е. либо бока, либо верх или низ должны касаться границ экрана
//    4. Применяя соответствующие маски сделайте так, чтобы когда устройство меняет ориентацию, то все клетки растягивались соответственно и ничего не вылетало за пределы экрана.

    self.field.autoresizingMask = UIViewAutoresizingFlexibleTopMargin |
                                  UIViewAutoresizingFlexibleBottomMargin |
                                  UIViewAutoresizingFlexibleLeftMargin |
                                  UIViewAutoresizingFlexibleRightMargin;
    
    //Мастер
    //5. При повороте устройства все черные клетки должны менять цвет :)
    //6.Сделайте так, чтобы доска при поворотах всегда строго находилась по центру
    
}

- (void)changeCellsColor:(UIColor*) color {
    for (UIView *view in self.field.subviews) {
        BOOL isCell = view.frame.size.height == self.cellSize;
        if (view.backgroundColor != [UIColor lightGrayColor] && isCell) {
            view.backgroundColor = color;
        }
    }
}

- (UIView *) randomChecker {
    return [self.checkersArray objectAtIndex:arc4random() % [self.checkersArray count]];
}

- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator {
    
    UIInterfaceOrientation interfaceOrientation = [[UIApplication sharedApplication] statusBarOrientation];
    
    if (UIInterfaceOrientationIsPortrait(interfaceOrientation)) {
        [self changeCellsColor:[UIColor greenColor]];
    } else if (UIInterfaceOrientationIsLandscape(interfaceOrientation)) {
        [self changeCellsColor:[UIColor blackColor]];
    }
    
    //8. Поставьте белые и красные шашки (квадратные вьюхи) так как они стоят на доске. Они должны быть сабвьюхами главной вьюхи (у них и у клеток один супервью)
    //9. После каждого переворота шашки должны быть перетасованы используя соответствующие методы иерархии UIView

    [UIView animateWithDuration:1 animations:^{
        NSInteger randomCountChange = arc4random() % ([self.checkersArray count] / 2) + 1;
        
        for (NSInteger i = 0; i < randomCountChange; i++) {
            UIView* firstChecker = [self randomChecker];
            UIView* secondChecker = [self randomChecker];
            
            CGRect tempRect = firstChecker.frame;
            [firstChecker setFrame:secondChecker.frame];
            [secondChecker setFrame:tempRect];
            
            [self.field bringSubviewToFront:firstChecker];
            [self.field bringSubviewToFront:secondChecker];
        }
    }];
}

@end

File: HW22 Touches/HW22/ViewController.swift
//
//  ViewController.swift
//  HW22
//
//  Created by Kirill Fedorov on 17/03/2019.
//  Copyright © 2019 Kirill Fedorov. All rights reserved.
//



//1. Создайте шахматное поле (8х8), используйте черные сабвьюхи
//2. Добавьте балые и красные шашки на черные клетки (используйте начальное расположение в шашках)
//3. Реализуйте механизм драг'н'дроп подобно тому, что я сделал в примере, но с условиями:
//4. Шашки должны ставать в центр черных клеток.
//5. Даже если я отпустил шашку над центром белой клетки - она должна переместиться в центр ближайшей к отпусканию черной клетки.
//6. Шашки не могут становиться друг на друга
//7. Шашки не могут быть поставлены за пределы поля.

import UIKit

class ViewController: UIViewController {
    
    var blackCells = [UIView]()
    var chessBoard: UIView!
    var checker = UIView()
    var touchOffSet = CGPoint()
    var checkers = [UIView]()
    
    func cellIsFree(cell: UIView, selectedChecker: UIView, chessBoard: UIView) -> Bool {
        for checker in checkers {
            if selectedChecker == checker {
                continue
            }
            if cell.point(inside: chessBoard.convert(checker.center, to: cell), with: nil) {
                return false
            }
        }
        return true
    }
    
    func findNearesBlackCell(blackCells: [UIView], checker: UIView, chessBoard: UIView) -> UIView {
        var delta = Float.greatestFiniteMagnitude
        var nearestCell: UIView!
        
        for blackCell in blackCells {
            let deltaX = fabsf(Float(blackCell.center.x - checker.center.x))
            let deltaY = fabsf(Float(blackCell.center.y - checker.center.y))
            
            if delta > sqrtf(Float(deltaX * deltaX + deltaY * deltaY)) &&
                cellIsFree(cell: blackCell, selectedChecker: checker, chessBoard: chessBoard) {
                nearestCell = blackCell
                delta = sqrtf(deltaX * deltaX + deltaY * deltaY)
            }
        }
        return nearestCell
    }

    fileprivate func configureChessBoard(_ cellsInLine: Int, _ cellSize: CGFloat, _ indentInCell: CGFloat, _ checkerSize: CGFloat) {
        let rangeWithOutCheckers = 3...4
        let blackCellImage = UIImage(named: "blackCell.jpg")
        for i in 0..<cellsInLine {
            for j in 0..<cellsInLine {
                guard i % 2 != j % 2 else { continue }
                let blackCell = UIView(frame: CGRect(x: cellSize * CGFloat(i),
                                                     y: cellSize * CGFloat(j),
                                                     width: cellSize,
                                                     height: cellSize))
                
                if let image = blackCellImage {
                    blackCell.backgroundColor = UIColor(patternImage: image)
                } else {
                    blackCell.backgroundColor = .black
                }
                
                chessBoard.addSubview(blackCell)
                blackCells.append(blackCell)
                
                if (j < rangeWithOutCheckers.lowerBound || j > rangeWithOutCheckers.upperBound) {
                    let checker = UIView.init(frame: CGRect(x: cellSize * CGFloat(i) + indentInCell,
                                                            y: cellSize * CGFloat(j) + indentInCell,
                                                            width: checkerSize,
                                                            height: checkerSize))
                    
                    checker.backgroundColor = j < rangeWithOutCheckers.lowerBound ? .white : .red
                    checker.layer.cornerRadius = checker.frame.width / 2
                    checker.tag = 1
                    chessBoard.addSubview(checker)
                    checkers.append(checker)
                }
            }
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        self.view.backgroundColor = .lightGray
        let minSide = min(self.view.frame.width, self.view.frame.height)
        let cellsInLine = 8
        let cellSize = minSide / CGFloat(cellsInLine)
        let multiplier = 0.7;
        let indentInCell = cellSize * CGFloat((1 - multiplier) / 2);
        let checkerSize = cellSize * CGFloat(multiplier)
        
        chessBoard = UIView(frame: CGRect(x: self.view.center.x - minSide / 2, y: self.view.center.y - minSide / 2, width: minSide, height: minSide))
        let chessBoardImage = UIImage(named: "chessBoard.jpg")
        if chessBoardImage != nil {
            chessBoard.backgroundColor = UIColor(patternImage: chessBoardImage!)
        }
        self.view.addSubview(chessBoard)
        
        chessBoard.autoresizingMask = [.flexibleRightMargin, .flexibleTopMargin, .flexibleBottomMargin, .flexibleLeftMargin]
        
        configureChessBoard(cellsInLine, cellSize, indentInCell, checkerSize)
    }
    
    // touches methods
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        
        guard let touch = touches.randomElement() else {
            return
        }
        let pointOnChessBoard = touch.location(in: chessBoard)
        let selectedView = chessBoard.hitTest(pointOnChessBoard, with: event)
        
        if let view = selectedView, view.tag == 1 {
            checker = view
            chessBoard.bringSubviewToFront(checker)
            
            let touchPoint = touch.location(in: checker)
            touchOffSet = CGPoint(x: (checker.bounds.midX) - (touchPoint.x),
                                  y: (checker.bounds.midY) - (touchPoint.y))
            
            UIView.animate(withDuration: 0.3) {
                self.checker.transform = self.checker.transform.scaledBy(x: 1.3, y: 1.3)
                self.checker.alpha = 0.7
            }
        }
    }
    
    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        guard let touch = touches.randomElement() else {
            return
        }
        let pointOnChessBoard = touch.location(in: chessBoard)
        let newPoint = CGPoint(x: pointOnChessBoard.x + touchOffSet.x, y: pointOnChessBoard.y + touchOffSet.y)
        self.checker.center = newPoint
    }
    
    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        let nearestCell = findNearesBlackCell(blackCells: blackCells, checker: checker, chessBoard: chessBoard)
        UIView.animate(withDuration: 0.3) {
            self.checker.center = nearestCell.center
            self.checker.transform = CGAffineTransform.identity
            self.checker.alpha = 1
        }
    }
    
    override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
        print(touches.count)
    }
}


